teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,612 @@
+
+<!-- ## Drawbacks -->
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+<!--
+What other approaches did you consider and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+### Efficiency
+#### Expression Language
+In Triggers, we use CEL for filtering to avoid spinning up a new pod. Similarly, we can choose a particular expression language - Common Expression Language - and use it to evaluate simple `Guards` efficiently in the controller. 
+```
+api: tekton.dev/v1alpha1
+kind: Pipeline
+metadata: 
+    name: generate-file
+spec:
+  params:
+    - name: branch
+      type: string
+  tasks: 
+    - name: create-file
+      taskRef: 
+        name: create-readme-file 
+    - name: check-file-exists 
+      when:
+        - name: should-echo
+          when: $(params.echo) == 'true'
+      taskRef:
+        name: echo-file-exists
+      runAfter:
+      - create-file
+```
+
+#### Expression Language Interceptor
+In Triggers, we use Common Expression Language for filtering to avoid spinning up a new pod using an interceptor which is pluggable. Similarly, we can provide a CEL interceptor out of the box and, if needed, users can add or bring their own interceptors to use other languages. 
+
+If needed, we can have a `Task` perform the logic needed and output a `Task` result, which the guarded `Tasks` can operate on using an expression language. The `Results` can be reused in multiple `Conditions` expressions thereby reducing the number of pods used. We can use a when property to specify the string expression to guardedly execute a `Task`. 
+
+### Skipping
+#### Dependency Type
+A `Task` can have two kinds of dependencies on another `Task`:
+- Resource dependency: based on resources needed from parent `Task`, which includes `Workspaces`, `Results` and `Resources`. 
+- Ordering dependency: based on `runAfter` which provides sequencing of `Tasks` when there may not be resource dependencies. 
+
+If we allow skipping of guarded `Tasks` and continuing with the dependent `Tasks`, there would be resolution errors in resource dependent `Tasks` since it needs resources from the guarded `Task` while the ordering dependent `Tasks` can run successfully. Thus, we can execute only the ordering dependent `Tasks` and terminate the resource dependent `Tasks`.
+
+#### Guard Location
+`Guards` can be defined in two locations:
+- Within a `Task`: if expression evaluates to `False`, only that `Task` is skipped. 
+- Between `Tasks`: if the expression evaluates to `False`, the subsequent `Tasks` are skipped.
+
+## Infrastructure Needed
+Separate project for the `CEL` controller and its test infrastructure - we'll create another proposal with more details.
+
+## Upgrade & Migration Strategy
```

Comment:
Migration Strategy - do we implement these (e.g. taskRef) on the current `conditions` field first - and later add an alias to `when` ?

I think that would be good - in case we want to change anything - we can omit adding it to the field named `when` - without waiting during the long deprecation period.